src/main/java/io/confluent/copycat/hdfs/HdfsWriter.java

```diff
@@ -1,167 +1,292 @@
+    }
+  }
+
+  public void recover(TopicPartition topicPart) {
     try {
-      RecordWriter<Long, SinkRecord> writer = getWriter(topicPart, record);
-      writer.write(System.currentTimeMillis(), record);
-      if (!offsets.containsKey(topicPart)) {
-        offsets.put(topicPart, record.getOffset() - 1);
-      }
-      updateRecordCounter(topicPart);
-      if (shouldRotate(topicPart)) {
-        rotate(topicPart);
+      switch (states.get(topicPart)) {
+        case RECOVERY_STARTED:
+          pause(topicPart);
+          nextState(topicPart);
+        case RECOVERY_PARTITION_PAUSED:
+          createWAL(topicPart);
+          nextState(topicPart);
+        case WAL_CREATED:
+          applyWAL(topicPart);
+          nextState(topicPart);
+        case WAL_APPLIED:
+          truncateWAL(topicPart);
+          nextState(topicPart);
+        case WAL_TRUNCATED:
+          resetOffsets(topicPart);
+          nextState(topicPart);
+        case OFFSET_RESET:
+          resume(topicPart);
+          nextState(topicPart);
       }
-    } catch (IOException e) {
-      throw new CopycatRuntimeException(e);
+    } catch (CopycatException e) {
+      log.error("Recovery failed.");
+      setRetryBackoff(backOffMs);
     }
   }
 
-  public void close() throws IOException, CopycatException {
-    for (TopicPartition topicPart: writers.keySet()) {
-      rotate(topicPart);
+  private void execute(TopicPartition topicPart) throws IOException {
+    Queue<SinkRecord> records = buffer.get(topicPart);
+    while(!records.isEmpty()) {
+      try {
+        switch (states.get(topicPart)) {
+          case RECOVERY_STARTED:
+            pause(topicPart);
+            nextState(topicPart);
+          case RECOVERY_PARTITION_PAUSED:
+            createWAL(topicPart);
+            nextState(topicPart);
+          case WAL_CREATED:
+            applyWAL(topicPart);
+            nextState(topicPart);
+          case WAL_APPLIED:
+            truncateWAL(topicPart);
+            nextState(topicPart);
+          case WAL_TRUNCATED:
+            resetOffsets(topicPart);
+            nextState(topicPart);
+          case OFFSET_RESET:
+            resume(topicPart);
+            nextState(topicPart);
+          case WRITE_STARTED:
+            pause(topicPart);
+            nextState(topicPart);
+          case WRITE_PARTITION_PAUSED:
+            writeRecord(topicPart, records.peek());
+            records.poll();
+            if (shouldRotate(topicPart)) {
+              nextState(topicPart);
+            } else {
+              break;
+            }
+          case SHOULD_ROTATE:
+            closeTempfile(topicPart);
+            nextState(topicPart);
+          case TEMP_FILE_CLOSED:
+            appendToWAL(topicPart);
+            nextState(topicPart);
+          case WAL_APPENDED:
+            commitFile(topicPart);
+            nextState(topicPart);
+          case FILE_COMMITTED:
+            setState(topicPart, State.WRITE_PARTITION_PAUSED);
```

Comment:
Sorry, maybe that wasn't clear.When we get to this FILE_COMMITTED state, we move on to WRITE_PARTITION_PAUSED. Normally the expectation for that state is that we have called `pause(topicPart)`. At this instant, that is true. But then if we immediately call `resume(topicPart)` as we do here, wouldn't that put us in the wrong state when we exit this loop and then get the next set of records? The topic partition will be resumed (active) and delivering messages, but the state we're in seems to assume it has been paused.

A similar way of thinking about this is, when we exit this loop/method, shouldn't we be in the same state as just after `recover()` as long as we haven't hit an error? The state we're in after `recover()` should be `WRITE_STARTED` since that's the one just after `OFFSET_RESET`, which is the last one we handle. Also, it intuitively makes sense since we want to be able to write data as long as we are in a good state.

Looking at this again, I think either the actions in the different states aren't quite right or we need add a bit. If we exit this after committing a file, I think we should reset to `WRITE_STARTED` and call `resume()`. This will ensure that if we've emptied the buffer, we can collect more records. (If we didn't empty the buffer, we should have kept on looping here anyway and exited via some other path.). Then, in `WRITE_STARTED` we can pause as we currently do. The difference, is that in `WRITE_PARTITION_PAUSED` we'd need to make sure to `resume()`  in the `else` block (and because of this, presumably reset to `WRITE_STARTED` state). Otherwise, under normal circumstances where we just weren't ready to rotate the file, wouldn't we block any more data from getting in (since we will have paused, break out of the switch statement, eventually exhaust the buffer and exit the loop, and never `resume()` that topic partition).

I'm actually not sure we need separate states for `WRITE_STARTED` and `WRITE_PARTITION_PAUSED`. Can the `pause(topicPart)` call even fail? If not, then I don't see why we'd need a separate step.

We might need a unit test that checks that multiple iterations of this process actually work properly.